resources/sdk/purecloudjava/templates/ApiClient.mustache

```diff
@@ -975,4 +1031,72 @@ public class ApiClient implements AutoCloseable {
         @Override
         public void close() throws Exception { }
     }
+
+    public static class Retry {
+        private final static long FIVE_MINUTES_IN_MILLIS = 300000L;
+        private final static int BASE_BACKOFF_TIME_IN_SEC = 3;
+        private final static int MAX_RETRIES_BEFORE_BACKOFF = 5;
+        private long totalRetryTimeInMillis = 0L;
+        private int retryCountBeforeBackOff = 0;
+        private long waitTimeInMillis = BASE_BACKOFF_TIME_IN_SEC * 1000L;
+        private Stopwatch stopwatch = null;
+        private long timeElapsedPerBucketInMillis = 0L;
+
+        private final List<Integer> statusCodes = Arrays.asList(429, 502, 503, 504);
+
+        public boolean shouldRetry(ApiClientConnectorResponse connectorResponse, long timeElapsedPerRequestInMillis) {
+            return totalRetryTimeInMillis < getMaxRetryTimeInMillis() && shouldWait(connectorResponse, timeElapsedPerRequestInMillis);
+        }
+
+        private boolean shouldWait(ApiClientConnectorResponse connectorResponse, long timeElapsedPerRequestInMillis) {
+            if (statusCodes.contains(connectorResponse.getStatusCode())) {
+
+                if (connectorResponse.getHeaders().containsKey("Retry-After")) {
+                    waitTimeInMillis = Integer.parseInt(connectorResponse.getHeaders().getOrDefault("Retry-After", String.valueOf(BASE_BACKOFF_TIME_IN_SEC))) * 1000L;
+                }
+                //If status code is 429 then wait until retry-after time and retry. OR If status code is retryable then for the first 5 times: wait until retry-after time and retry.
+                if (connectorResponse.getStatusCode() == 429 || retryCountBeforeBackOff++ < MAX_RETRIES_BEFORE_BACKOFF) {
+                    totalRetryTimeInMillis += (waitTimeInMillis + timeElapsedPerRequestInMillis);
+                    waitBeforeRetry(waitTimeInMillis);
+                    return true;
+                }
+                //If status code is 50x and if 3 Sec and 9 Sec buckets were already retried then wait for every 27 Sec and retry.
+                if (timeElapsedPerBucketInMillis > (FIVE_MINUTES_IN_MILLIS * 2)) {
+                    waitTimeInMillis = getWaitTimeExp(3);
+                    totalRetryTimeInMillis += (waitTimeInMillis + timeElapsedPerRequestInMillis);
+                    waitBeforeRetry(waitTimeInMillis);
+                    return true;
```

Comment:
A lot of the code in this class can be streamlined. For example, these four can be reduced to one by making waitBeforeRetry return true and replacing these lines with: `return waitBeforeRetry(getWaitTimeExp(3))`. Tracking the total time via math is unnecessary as mentioned abvove.